### PR TITLE
Fix #545: Low fi onboarding part 1

### LINKED ISSUE #545: UI: Onboarding flow
This issue is tracking onboarding flow for new learners. 
https://xd.adobe.com/spec/e2239cf4-9cde-4c08-5296-25316c1f0a14-9412/screen/5981ecd6-2a80-4d82-8715-3e92294d53b5/RG-OB-Onboarding-1-

### PR DESCRIPTION
**Pull request by rt4914**
## Explanation

Mock: https://xd.adobe.com/spec/e2239cf4-9cde-4c08-5296-25316c1f0a14-9412/screen/5981ecd6-2a80-4d82-8715-3e92294d53b5/RG-OB-Onboarding-1-

This PR implements the onboarding flow for new learners.

onboarding_flow

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+  @Test
+  fun testOnboardingFragment_checkSlide3Image_isCorrect() {
+    launch(OnboardingActivity::class.java).use {
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(
+        withIndex(
+          withId(R.id.slide_image_view),
+          2
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: @BenHenning I am using one function `withIndex` which has been defined below. It helps to check the content for a specific item in ViewPager. But the issue is that the it will keep index only relative to the stack. For example: if there are 10 items in viewpager and you are going to last item of viewpager, which is position 9, the index for testing with this function will most probably be 2 or 3 considering that viewpager will have only that many loaded in its stack. So any suggestions on how can I fix this?

**Comment by BenHenning** [comment]
I didn't review it in full yet, but can you first upload a recording of the new flow? It'd be good to see it context to reference it as part of the review.

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt (hunk 1 above)
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: This index thing seems a bit hacky, and the issue you describe seems difficult to solve robustly. Could we instead leverage tags. I'm thinking we tag each view pager view with information corresponding to the tab that view represents, then we can match based on the tag. It doesn't solve the needing to swipe bit, but we could conceivably implement a swipeToTab mechanism if we wanted to (though I think what you have now for swiping seems reasonable).

**Comment by BenHenning** [comment]
@rt4914 feel free to follow up with me via chat if you want to discuss the solution proposed here in more detail.

**Comment by rt4914** [comment]
> @rt4914 feel free to follow up with me via chat if you want to discuss the solution proposed here in more detail.

@BenHenning I have just updated the PR description to include gif. But let's wait for for @veena14cs and @nikitamarysolomanpvt 's review and meanwhile I will try to check your comment and apply accordingly.

File: app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        title.set(context.resources.getString(R.string.slide_2_title))
+        description.set(context.resources.getString(R.string.slide_2_description))
+      }
+      3 -> {
+        slideImage.set(R.drawable.ic_onboarding_3)
+        title.set(context.resources.getString(R.string.slide_3_title))
+        description.set(context.resources.getString(R.string.slide_3_description))
+      }
+      else -> {
+        slideImage.set(R.drawable.ic_onboarding_0)
```
Comment (veena14cs) [inline app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt]: Please check if this image id is right here.

File: app/src/main/res/layout/onboarding_fragment.xml [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        android:id="@+id/slide_dots_container"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:orientation="horizontal"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent">
+
+        <View
```
Comment (veena14cs) [inline app/src/main/res/layout/onboarding_fragment.xml]: Can this be made dynamic. If suppose onboarding screens are increased or decreased depending on any of the new feature, then we will have to add these views accordingly. Just a suggestion.

**Review by veena14cs — APPROVED** [review]
LGTM besides PTAL at some comments.

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt (hunk 1 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: @BenHenning In your suggestion, actually it would be similar to test the tag or text or anything for a view/viewgroup. If I am correct, the main reason behind your suggestion was to remove the dependency on index. So I finally removed the dependency on index and used below code which also, does not use tags.

```
@Test
  fun testOnboardingFragment_checkSlide3Image_isCorrect() {
    launch(OnboardingActivity::class.java).use {
      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
      onView(
        allOf(
          withId(R.id.slide_image_view),
          isCompletelyDisplayed()
        )
      ).check(matches(withTagValue(equalTo(R.drawable.ic_onboarding_3))))
    }
  }
```

you can see that I have used `isCompletelyDisplayed` which is need to make sure that the current view is actually visible on screen and is not in stack. The main issue that I am facing is that, because of animation and the time needed for swipe, the test case is failing. Which means that we need some sort of waiting mechanism. 

I took reference from `HomeActivityTest` and also https://stackoverflow.com/a/22563297/3689782 but none of these solutions seems to work. Also, if I put `SystemClock.sleep(1000L)` it does work perfectly. So can you suggest any suggestion on how can we fix this.

Also, in your HomeActivityTest, you have written code to wait for [30000L](https://github.com/oppia/oppia-android/blob/e3c6553e925c91c9c7860bd29f23c1e5de9d437a/app/src/sharedTest/java/org/oppia/app/home/HomeActivityTest.kt#L358) but when we run the app, it does work correctly but does not take 30 sec. So how exactly is it working.

In short, if we use `SystemClock.sleep(500)` it does work, its just that we need proper alternative of this.

File: app/src/main/res/layout/onboarding_fragment.xml (hunk 3 above)
Comment (rt4914) [reply app/src/main/res/layout/onboarding_fragment.xml]: @veena14cs Nice point. I will update the implementation accordingly.

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+/** Tests for [OnboardingFragment]. */
+@RunWith(AndroidJUnit4::class)
+class OnboardingFragmentTest {
+
+  @Before
+  fun setUp() {
+    Intents.init()
+  }
+
+  @Test
```
Comment (nikitamarysolomanpvt) [inline app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt:43]: Please add test cases for indicator.

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+    launch(OnboardingActivity::class.java).use {
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(
+        withIndex(
+          withId(R.id.slide_description_text_view),
+          2
+        )
+      ).check(matches(withText(getResources().getString(R.string.slide_3_description))))
```
Comment (nikitamarysolomanpvt) [inline app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: ditto: You may directly use withText(final int resourceId) instead of getting string from resource.
Please follow this through out.

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+    launch(OnboardingActivity::class.java).use {
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(
+        withIndex(
+          withId(R.id.slide_title_text_view),
+          2
+        )
+      ).check(matches(withText(getResources().getString(R.string.slide_3_title))))
```
Comment (nikitamarysolomanpvt) [inline app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: You may directly use withText(final int resourceId) instead of getting string from resource.

**Review by nikitamarysolomanpvt — COMMENTED** [review]
I also think if sliders and indicators are dynamic would be better,Please see the comments and assign me back once new implementation changes are done

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt (hunk 5 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: Done. Thanks.

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt (hunk 6 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: Done. Thanks.

File: app/src/main/res/layout/onboarding_fragment.xml (hunk 3 above)
Comment (rt4914) [reply app/src/main/res/layout/onboarding_fragment.xml]: Done. Thanks.

File: app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt (hunk 2 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt]: Yes it was incorrect. Updated this to last slide image.

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt (hunk 4 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt:43]: Added new test cases along with new implementation for dot indicator.

**Comment by rt4914** [comment]
@veena14cs @nikitamarysolomanpvt I have resolved all your comments and made changes accordingly. PTAL.

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt (hunk 1 above)
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: That HomeActivityTest mechanism waits up to 30 seconds for an action to occur and otherwise fails. It’s a bit friendlier to Espresso since it yields time for idling to properly occur. 

We should be disabling animations on emulators and devices when running tests. Is this actually an animation issue or a data loading/rendering issue? With animations disabled and onIdle satisfied, I’d expect this to pass unless we have coroutines somewhere blocking the UI (in which case this devolves to #89).

Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: Actually, animations are disabled but then also the sliding of viewpager and because of which these test cases are failing. Also, I did try your `HomeActivityTest` along with animation off but that did not work for me. I will try to investigate on this more today before confirming that whether it is needs #89 or not.

File: app/src/main/java/org/oppia/app/splash/SplashActivity.kt [hunk 7]
```diff
@@ -14,8 +15,9 @@ class SplashActivity : AppCompatActivity() {
     super.onCreate(savedInstanceState)
     setContentView(R.layout.splash_activity)
 
-    window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
-    val intent = Intent(this@SplashActivity, ProfileActivity::class.java)
+    window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN)
+    // TODO (#565): Control the next activity based on whether the leaner is using the app for the first time or not.
```
Comment (veena14cs) [inline app/src/main/java/org/oppia/app/splash/SplashActivity.kt]: Typo in Todo comment `learner`.

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt (hunk 1 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: @BenHenning I have created this https://gist.github.com/rt4914/17ef7c237b34fb4b5186080b14a78f16
which contains code that I have tried along with error and description.

**Review by veena14cs — APPROVED** [review]
LGTM

File: app/src/main/java/org/oppia/app/splash/SplashActivity.kt (hunk 7 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/splash/SplashActivity.kt]: Resolved.

**Review by nikitamarysolomanpvt — APPROVED** [review]
LGTM

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt (hunk 1 above)
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: When you watch one of the failing tests run, what do you see? Is it actually swiping to the right tab, then hanging as the 30 second time waits to find the element but then fails despite the element actually being on the screen? Does it not swipe? Does it fail immediately after moving to the new tab?

Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: @BenHenning In the emulator it actually follows all the steps except waiting for 30 secs. It is working as if the 30 sec thing is not even working. (Same thing happens when i run HomeActivityTest). Also, all the steps in the emulator are actually working correctly but in the end test case fails with error: `The view was not visible 100%` because we are using `isCompletelyDisplayed()`.

Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: If it’s not waiting that means the condition that’s waiting 30 seconds is passing, which is presumably the withId. So it’s finding the element within 30 seconds, but something about the screen is causing isCompletelyDisplayed to fail. I’m not sure why that’s failing without investigating it in detail myself.

File: app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+class OnboardingSlideViewModel(val context: Context, val index: Int) : ObservableViewModel() {
+  val slideImage = ObservableField<Int>(R.drawable.ic_onboarding_0)
+  val title = ObservableField<String>(context.resources.getString(R.string.slide_0_title))
+  val description = ObservableField<String>(context.resources.getString(R.string.slide_0_description))
+
+  init {
+    slideChanged(index)
+  }
+
+  private fun slideChanged(index: Int) {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt]: This method name is a bit misleading: the index is immutable, so this is really initializing state not changing it.

File: app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import android.content.Context
+import androidx.databinding.ObservableField
+import androidx.lifecycle.ViewModel
+import org.oppia.app.R
+import org.oppia.app.viewmodel.ObservableViewModel
+
+const val TOTAL_NUMBER_OF_SLIDES = 4
+
+/** [ViewModel] for slide in onboarding flow. */
+class OnboardingSlideViewModel(val context: Context, val index: Int) : ObservableViewModel() {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt]: Can we use an enum instead of an int index? Prefer using strongly typed values instead of overloading ints/strings. That tends to be less error prone which helps avoid pitfalls when others change this code in the future.

File: app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+  private val activity: AppCompatActivity,
+  private val fragment: Fragment,
+  private val viewModelProvider: ViewModelProvider<OnboardingViewModel>
+) {
+  private val dotsList = ArrayList<ImageView>()
+  private lateinit var onboardingPagerAdapter: OnboardingPagerAdapter
+  private val routeToProfileListener = activity as RouteToProfileListener
+  private lateinit var binding: OnboardingFragmentBinding
+  private lateinit var slidesViewPager: ViewPager
+  fun handleCreateView(inflater: LayoutInflater, container: ViewGroup?): View? {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt:31]: Nit: methods should always have empty spacing around them to make them easier to identify at a glance. There should be an empty newline above this line.

File: app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+      }
+    })
+  }
+
+  fun clickOnGetStarted() {
+    routeToProfileListener.routeToProfile()
+  }
+
+  fun clickOnSkip() {
+    getOnboardingViewModel().slideChanged(3)
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt]: This would also benefit from switching to an enum since it will become more readable. As it stands, ‘3’ is a magic value with no context in this method, and it’s fragile in the sense that changes to the onboarding stages could break this code.

File: app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    val dotIdList = ArrayList<Int>()
+    dotIdList.add(R.id.onboarding_dot_0)
+    dotIdList.add(R.id.onboarding_dot_1)
+    dotIdList.add(R.id.onboarding_dot_2)
+    dotIdList.add(R.id.onboarding_dot_3)
+    for (index in 0 until TOTAL_NUMBER_OF_SLIDES) {
+      val dotView = ImageView(activity)
+      dotView.id = dotIdList[index]
+      dotView.setImageResource(R.drawable.onboarding_dot_active)
+      if (index != 0) {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt]: This seems a bit obscure to me. Perhaps it would be cleaner to omit this and call selectDot at the end of this method to select the initial page?

File: app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+  fun handleCreateView(inflater: LayoutInflater, container: ViewGroup?): View? {
+    binding = OnboardingFragmentBinding.inflate(inflater, container, /* attachToRoot= */ false)
+    // NB: Both the view model and lifecycle owner must be set in order to correctly bind LiveData elements to
+    // data-bound view models.
+    binding.let {
+      it.lifecycleOwner = fragment
+      it.presenter = this
+      it.viewModel = getOnboardingViewModel()
+    }
+    slidesViewPager = binding.root.onboarding_slide_view_pager as ViewPager
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt]: It seems a bit cleaner to use the direct access variables, eg ``binding.onboardingSlideViewPager``. This avoids needing the synthetic imports.

File: app/src/main/java/org/oppia/app/onboarding/OnboardingPagerAdapter.kt [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    container.addView(binding.root)
+    return binding.root
+  }
+
+  override fun getCount(): Int {
+    return TOTAL_NUMBER_OF_SLIDES
+  }
+
+  override fun isViewFromObject(view: View, `object`: Any): Boolean {
+    return view == `object`
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/onboarding/OnboardingPagerAdapter.kt:26]: Can you explain this logic? I’m not sure I follow.

File: app/src/main/res/values/strings.xml [hunk 15]
```diff
@@ -138,6 +138,17 @@
   <string name="add_profile_error_name_only_letters">Names can only have letters. Try another name?</string>
   <string name="add_profile_error_pin_length">Your PIN should be 3 digits long.</string>
   <string name="add_profile_error_pin_confirm_wrong">Please make sure that both PINs match.</string>
+  <!-- OnboardingActivity -->
+  <string name="slide_0_title">Welcome to Oppia!</string>
```
Comment (BenHenning) [inline app/src/main/res/values/strings.xml]: Here and below: suggest including ‘onboarding’ in the names to make their context clearer.

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt [hunk 16]
```diff
@@ -0,0 +1,486 @@
...
+    launch(OnboardingActivity::class.java).use {
+      onView(withId(R.id.skip_text_view)).check(matches(isDisplayed()))
+    }
+  }
+
+  @Test
+  fun testOnboardingFragment_checkDefaultSlide_clickSkipButton_shiftsToLastSlide() {
+    launch(OnboardingActivity::class.java).use {
+      onView(withId(R.id.skip_text_view)).perform(click())
+      SystemClock.sleep(500)
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: Here and elsewhere: we should use our custom waitFor() instead of a hard sleep. Sleeps like this are fragile and slow down the tests.

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt [hunk 17]
```diff
@@ -0,0 +1,486 @@
...
+      onView(
+        allOf(
+          withId(R.id.slide_title_text_view),
+          isCompletelyDisplayed()
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt:123]: Why again do we need to check that they are completely displayed?

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt [hunk 18]
```diff
@@ -0,0 +1,486 @@
...
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      onView(withId(R.id.onboarding_slide_view_pager)).perform(swipeLeft())
+      SystemClock.sleep(500)
+      onView(withId(R.id.get_started_button)).perform(click())
+      intended(hasComponent(ProfileActivity::class.java.name))
+    }
+  }
+
+  @Test
+  fun testOnboardingFragment_swipeLeftOnLastSlide_isNotWorking() {
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: Here and for similarly named tests, I think the outcome should be something like ‘doesNotWork’ rather than is not working. I think that might read a bit better.

File: app/src/main/java/org/oppia/app/onboarding/RouteToProfileListener.kt [hunk 19]
```diff
@@ -0,0 +1,6 @@
+package org.oppia.app.onboarding
+
+/** Listener for when an activity should route to [ProfileActivity]. */
+interface RouteToProfileListener {
+  fun routeToProfile()
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/onboarding/RouteToProfileListener.kt]: This method name is ambiguous. I think you meant ‘routeToProfileList’ not to a specific profile.

**Review by BenHenning — COMMENTED** [review]
Thanks! Some of my comments may be redundant, but I did a full review this time.

File: app/src/main/java/org/oppia/app/onboarding/RouteToProfileListener.kt (hunk 19 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/onboarding/RouteToProfileListener.kt]: Update interface and function name to following:

```
interface RouteToProfileListListener {
  fun routeToProfileList()
}
```

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt (hunk 18 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: Updated as per comment.

File: app/src/main/res/values/strings.xml (hunk 15 above)
Comment (rt4914) [reply app/src/main/res/values/strings.xml]: Done here and everywhere else.

File: app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt (hunk 13 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt]: Done. removed `slidesViewPager`

File: app/src/main/java/org/oppia/app/onboarding/OnboardingPagerAdapter.kt (hunk 14 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/onboarding/OnboardingPagerAdapter.kt:26]: The method `instantiateItem(ViewGroup, int)` returns Object for a particular view. PagerAdapter implementation is considering this Object as a key value when viewpager changes a page.

So, if we return the view itself from `instantiateItem(ViewGroup, int)`, then our key for that page becomes the view itself. We can check `return view == object;` from `isViewFromObject (View view, Object object)` which will always return true and our pages will display


@BenHenning Actually, this was a part of PagerAdapter extension and some google searches.

Mainly you can find the detail description here: https://stackoverflow.com/a/32709553

File: app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt (hunk 12 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt]: Yeah, that makes sense actually. Removed this if-else and after for loop, called `selectDot(0)`

File: app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt (hunk 10 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt:31]: Done.

File: app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt (hunk 9 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt]: Introduced `ViewPagerSlide` as enum. I think it will have nit changes with names.

File: app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt (hunk 11 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt]: Changed after using enum.

File: app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt (hunk 8 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/onboarding/OnboadingSlideViewModel.kt]: Changed to `initializingSlide`

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt (hunk 17 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt:123]: `isDisplayed` return true even in if the view is in background and is not visible completely on the screen, whereas `isCompletelyDisplayed` makes sure that the view is 100% visible on the screen.

I can give you more information on this with correct test cases once this animation related issue is resolved because currently it is hard to give a perfect example for the above question as test cases are failing because of this waiting thing also.

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt (hunk 16 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: Minimal code with only one test case which uses waitFor() code, currently this test case is failing but in mobile is working as we are expecting it to be.

https://gist.github.com/rt4914/d4793eb6e7be36615e75966e14c36ff3

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt (hunk 1 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: @BenHenning I tried various things on this, mainly using the 30sec wait, but that was also not working. As a result I believe that you will need to investigate this issue.

File: app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+      it.viewModel = getOnboardingViewModel()
+    }
+    setUpViewPager()
+    addDots()
+    return binding.root
+  }
+
+  private fun setUpViewPager() {
+    onboardingPagerAdapter = OnboardingPagerAdapter(fragment.requireContext())
+    binding.onboardingSlideViewPager.offscreenPageLimit = TOTAL_NUMBER_OF_SLIDES - 1
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt]: @BenHenning This line of code is only needed to make sure that test cases pass otherwise we do not need this code. So considering that, is this correct approach, because I think that test cases should not introduced any production code.

Anyways, because of this line, now we can use items of viewpager as indexes similar to recyclerview.
I have introduced a custom matcher in test cases `withIndex` which helps us to match items at a particular indexes. https://github.com/oppia/oppia-android/blob/4f7861e5179f72e385dec44ac8a6f4e648500ef2/app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt#L179

**Review by rt4914 — COMMENTED** [review]
@BenHenning PTAL

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt (hunk 1 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: Two major points on how all these test cases have been fixed:

1. As per @BenHenning 's suggestion I replaced `swipe` with `ViewPagerActions.swipeToPage()` and that fixed some of the issues. `swipeToPage()` waits for everything to finish before going to the next step in test case.

2. Using `show-layout-bounds` from developer options in mobile, I was able to see some boxes which were actually meant to be `Visibility.GONE`, so that was also fixed, but changing the margin to padding.

File: app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt (hunk 16 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt]: Two major points on how all these test cases have been fixed:

1. As per @BenHenning 's suggestion I replaced `swipe` with `ViewPagerActions.swipeToPage()` and that fixed some of the issues. `swipeToPage()` waits for everything to finish before going to the next step in test case.

2. Using `show-layout-bounds` from developer options in mobile, I was able to see some boxes which were actually meant to be `Visibility.GONE`, so that was also fixed, but changing the margin to padding.

File: app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt (hunk 20 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt]: Two major points on how all these test cases have been fixed:

1. As per @BenHenning 's suggestion I replaced `swipe` with `ViewPagerActions.swipeToPage()` and that fixed some of the issues. `swipeToPage()` waits for everything to finish before going to the next step in test case.

2. Using `show-layout-bounds` from developer options in mobile, I was able to see some boxes which were actually meant to be `Visibility.GONE`, so that was also fixed, but changing the margin to padding.

**Comment by rt4914** [comment]
@veena14cs @nikitamarysolomanpvt You have already reviewed this PR and approved it. But there have been changes to test cases. PTAL so that I can pass it to Ben afterwards.

**Review by veena14cs — APPROVED** [review]
LGTM.

**Review by nikitamarysolomanpvt — APPROVED** [review]
LGTM

File: app/src/main/res/layout/splash_activity.xml [hunk 21]
```diff
@@ -5,4 +5,3 @@
   android:layout_width="match_parent"
   android:layout_height="match_parent"
   tools:context=".splash.SplashActivity"/>
```
Comment (nikitamarysolomanpvt) [inline app/src/main/res/layout/splash_activity.xml]: please revert this change.

Comment (rt4914) [reply app/src/main/res/layout/splash_activity.xml]: This was done along with changing test case in this PR. So i think it is fine for this PR that keep this test cases.

**Comment by rt4914** [comment]
@veena14cs @nikitamarysolomanpvt  for testing this, you will need to open the app, finish the onboarding flow, close the app completely and then again open it, it should show profile activity in second attempt.

You need to close the app completely because of issue number #322

File: app/src/sharedTest/java/org/oppia/app/splash/SplashActivityTest.kt [hunk 22]
```diff
@@ -26,19 +67,140 @@ class SplashActivityTest {
...
+      intended(hasComponent(ProfileActivity::class.java.name))
+    }
   }
 
-  @After
-  fun tearDown() {
-    Intents.release()
+  private fun simulateNewAppInstance() {
+    // Simulate a fresh app install by clearing any potential on-disk caches using an isolated onboarding flow controller.
+    createTestRootComponent().getOnboardingFlowController().clearOnboardingFlow()
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/splash/SplashActivityTest.kt]: @BenHenning In this `clearOnboardingFlow` cannot be used now as it was removed from @nikitamarysolomanpvt 's PR as per your comments. Now, the splash screen works in following ways:
1. if the user has already onboarded -> open `ProfileActivity` else open `OnboardingFlowActivity`.

Now to test this correctly, we need to clear the cache of onboarding flow, but that function has been removed from the OnboardingFlowController file. So how can I still do that cache-clear thing in app module test cases.

Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/splash/SplashActivityTest.kt]: Can we write something similar to ProfileTestHelper in which there can be a function which helps in clearing this cache?

Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/app/splash/SplashActivityTest.kt]: Yes, I think a test helper is a good way to solve this problem rather than exposing a test-only API in the domain controller. In such a helper, you could either:

1) Delete the actual file from the filesystem

2) Create a new persistent cache store object and use its clear method

In either case, you should move the name of the cache to a dagger constant so that you can share & inject it both in the domain controller and test helper. I suggest doing this as a dedicated PR to keep things clean.

Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/splash/SplashActivityTest.kt]: Created PR #670 for this.

Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/splash/SplashActivityTest.kt]: Done. via just recreating the root component in test file
